apiserver/facades/client/application/application.go

```diff
@@ -2622,3 +2646,227 @@ func validateAgentVersions(application Application, versioner AgentVersioner) er
 	}
 	return nil
 }
+
+// UnitsInfo isn't on the v11 API.
+func (u *APIv11) UnitsInfo(_, _ struct{}) {}
+
+// UnitsInfo returns unit information.
+func (api *APIBase) UnitsInfo(in params.Entities) (params.UnitInfoResults, error) {
+	out := make([]params.UnitInfoResult, len(in.Entities))
+	leaders, err := api.leadershipReader.Leaders()
+	if err != nil {
+		return params.UnitInfoResults{}, errors.Trace(err)
+	}
+	for i, one := range in.Entities {
+		tag, err := names.ParseUnitTag(one.Tag)
+		if err != nil {
+			out[i].Error = common.ServerError(err)
+			continue
+		}
+		unit, err := api.backend.Unit(tag.Id())
+		if err != nil {
+			out[i].Error = common.ServerError(err)
+			continue
+		}
+		app, err := api.backend.Application(unit.ApplicationName())
+		if err != nil {
+			out[i].Error = common.ServerError(err)
+			continue
+		}
+		curl, _ := app.CharmURL()
+		machineId, _ := unit.AssignedMachineId()
+		workloadVersion, err := unit.WorkloadVersion()
+		if err != nil {
+			out[i].Error = common.ServerError(err)
+			continue
+		}
+
+		result := &params.UnitResult{
+			Tag:             tag.String(),
+			WorkloadVersion: workloadVersion,
+			Machine:         machineId,
+			Charm:           curl.String(),
+		}
+		if leader := leaders[unit.ApplicationName()]; leader == unit.Name() {
+			result.Leader = true
+		}
+		if machineId != "" {
+			machine, err := api.backend.Machine(machineId)
+			if err != nil {
+				out[i].Error = common.ServerError(err)
+				continue
+			}
+			publicAddress, err := machine.PublicAddress()
+			if err == nil {
+				result.PublicAddress = publicAddress.Value
+			}
+			openPorts, err := api.openPortsOnMachine(unit.Name(), machineId)
+			if err != nil {
+				out[i].Error = common.ServerError(err)
+				continue
+			}
+			result.OpenedPorts = openPorts
+		}
+		container, err := unit.ContainerInfo()
+		if err != nil && !errors.IsNotFound(err) {
+			out[i].Error = common.ServerError(err)
+			continue
+		}
+		if err == nil {
+			if addr := container.Address(); addr != nil {
+				result.Address = addr.Value
+			}
+			result.ProviderId = container.ProviderId()
+			if len(result.OpenedPorts) == 0 {
+				result.OpenedPorts = container.Ports()
+			}
+		}
+		result.RelationData, err = api.relationData(app, unit)
+		if err != nil {
+			out[i].Error = common.ServerError(err)
+			continue
+		}
+
+		out[i].Result = result
+	}
+	return params.UnitInfoResults{out}, nil
+}
+
+func (api *APIBase) openPortsOnMachine(unitName, machineID string) ([]string, error) {
+	var result []string
+	allOpenPorts, err := api.model.AllPorts()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// We are only looking for the open ports on the "" subnet.
```

Comment:
Why are we only looking for open ports on the "" subnet?

Reply:
That's what status does - the same logic was applied here. Not 100% sure why